crates/bevy_ecs/src/query/fetch.rs

```diff
@@ -49,6 +54,213 @@ pub trait WorldQuery {
 
 pub type QueryItem<'w, 's, Q> = <<Q as WorldQuery>::Fetch as Fetch<'w, 's>>::Item;
 
+/// Types that implement this trait are responsible for fetching query items from tables or
+/// archetypes.
+///
+/// Every type that implements [`WorldQuery`] have their associated [`WorldQuery::Fetch`] and
+/// [`WorldQuery::State`] types that are essential for fetching component data. If you want to
+/// implement a custom query type, you'll need to implement [`Fetch`] and [`FetchState`] for
+/// those associated types.
+///
+/// You may want to implement a custom query for the following reasons:
+/// - Named structs can be clearer and easier to use than complex query tuples. Access via struct
+///   fields is more convenient than destructuring tuples or accessing them via `q.0, q.1, ...`
+///   pattern and saves a lot of maintenance burden when adding or removing components.
+/// - Nested queries enable the composition pattern and makes query types easier to re-use.
+/// - You can bypass the limit of 15 components that exists for query tuples.
+///
+/// Implementing the trait manually can allow for a fundamentally new type of behaviour.
+///
+/// # Derive
+///
+/// This trait can be derived with the [`derive@super::Fetch`] macro.
+/// To do so, all fields in the struct must themselves impl [`WorldQuery`].
+///
+/// The derive macro implements [`WorldQuery`] for your type and declares two structs that
+/// implement [`Fetch`] and [`FetchState`] and are used as [`WorldQuery::Fetch`] and
+/// [`WorldQuery::State`] associated types respectively.
+///
+/// **Note:** currently, the macro only supports named structs.
```

Comment:
I'm surprised by this limitation: tuple structs *are* named structs, just with dumb field names of `0`, `1` etc.

See https://github.com/rust-lang/rust-clippy/issues/7985

Reply:
Oh, cool, I forgot about this. This indeed makes it a lot easier to support tuple structs

Reply:
I've just tried it - unfortunately, it doesn't make it a lot easier to implement a macro.

Tuple structs can't be declared like this:
```
struct Test {
    0: i32,
}
```

Also, `Ident::new("0", Span::call_site());` panics.

Without this, adding tuple structs support will be more difficult. Btw, our `Bundle` derive doesn't support tuple structs either, so I'd prefer to leave it as is for now for consistency and simplicity sake.

Reply:
This is fine by me for now. Although do note that #3634 exists.